library/md.c

```diff
@@ -228,6 +237,145 @@ const mbedtls_md_info_t *mbedtls_md_info_from_type(mbedtls_md_type_t md_type)
     }
 }
 
+#if defined(MBEDTLS_PSA_CRYPTO_C)
+psa_algorithm_t mbedtls_md_psa_alg_from_type(mbedtls_md_type_t md_type)
+{
+    switch (md_type) {
+#if defined(MBEDTLS_MD_CAN_MD5)
+        case MBEDTLS_MD_MD5:
+            return PSA_ALG_MD5;
+#endif
+#if defined(MBEDTLS_MD_CAN_RIPEMD160)
+        case MBEDTLS_MD_RIPEMD160:
+            return PSA_ALG_RIPEMD160;
+#endif
+#if defined(MBEDTLS_MD_CAN_SHA1)
+        case MBEDTLS_MD_SHA1:
+            return PSA_ALG_SHA_1;
+#endif
+#if defined(MBEDTLS_MD_CAN_SHA224)
+        case MBEDTLS_MD_SHA224:
+            return PSA_ALG_SHA_224;
+#endif
+#if defined(MBEDTLS_MD_CAN_SHA256)
+        case MBEDTLS_MD_SHA256:
+            return PSA_ALG_SHA_256;
+#endif
+#if defined(MBEDTLS_MD_CAN_SHA384)
+        case MBEDTLS_MD_SHA384:
+            return PSA_ALG_SHA_384;
+#endif
+#if defined(MBEDTLS_MD_CAN_SHA512)
+        case MBEDTLS_MD_SHA512:
+            return PSA_ALG_SHA_512;
+#endif
+        default:
+            return PSA_ALG_NONE;
+    }
+}
+
+mbedtls_md_type_t mbedtls_md_type_from_psa_alg(psa_algorithm_t psa_alg)
+{
+    switch (psa_alg) {
+#if defined(MBEDTLS_MD_CAN_MD5)
+        case PSA_ALG_MD5:
+            return MBEDTLS_MD_MD5;
+#endif
+#if defined(MBEDTLS_MD_CAN_RIPEMD160)
+        case PSA_ALG_RIPEMD160:
+            return MBEDTLS_MD_RIPEMD160;
+#endif
+#if defined(MBEDTLS_MD_CAN_SHA1)
+        case PSA_ALG_SHA_1:
+            return MBEDTLS_MD_SHA1;
+#endif
+#if defined(MBEDTLS_MD_CAN_SHA224)
+        case PSA_ALG_SHA_224:
+            return MBEDTLS_MD_SHA224;
+#endif
+#if defined(MBEDTLS_MD_CAN_SHA256)
+        case PSA_ALG_SHA_256:
+            return MBEDTLS_MD_SHA256;
```

Comment:
What if `MBEDTLS_PSA_ACCEL_ALG_SHA_256` will be defined, but not `MBEDTLS_SHA256_C`? 
I'm looking for a reason that there's a `MBEDTLS_MD_SHA256_VIA_PSA`, but not `MBEDTLS_MD_SHA256_VIA_LEGACY`.

Reply:
> What if `MBEDTLS_PSA_ACCEL_ALG_SHA_256` will be defined, but not `MBEDTLS_SHA256_C`?

That's precisely the case we want to support: in that case, SHA-256 will be dispatched to PSA which will call the appropriate driver.

> I'm looking for a reason that there's a `MBEDTLS_MD_SHA256_VIA_PSA`, but not `MBEDTLS_MD_SHA256_VIA_LEGACY`.

I don't think we need two macros, because the information is essentially one boolean: we dispatch either to PSA or to legacy. So, `xxx_VIA_LEGACY` would be defined if and only if `xxx_VIA_PSA` isn't.

Note that dispatching to legacy can result in returning `FEATURE_UNAVAILABLE` in case the corresponding hash is not defined in legacy. But that's already the case, and always has been.